lib/Command/RecreateMasterKey.php

```diff
@@ -0,0 +1,326 @@
+		$this->userSetup = $userSetup;
+		$this->mailer = $mailer;
+		$this->secureRandom = $secureRandom;
+		$this->l = $l;
+	}
+
+	protected function configure() {
+		parent::configure();
+
+		$this
+			->setName('encryption:recreate-master-key')
+			->setDescription('Replace existing master key with new one. Encrypt the file system with newly created master key')
+		;
+
+		$this->addOption(
+			'yes',
+			'y',
+			InputOption::VALUE_NONE,
+			'Answer yes to all questions'
+		);
+	}
+
+	protected function execute(InputInterface $input, OutputInterface $output) {
+		$yes = $input->getOption('yes');
+		if ($this->util->isMasterKeyEnabled()) {
+			$question = new ConfirmationQuestion(
+				'Warning: Inorder to re-create master key, the entire ownCloud filesystem will be decrypted and then encrypted using new master key.'
+				. ' Do you want to continue? (y/n)', false);
+			if ($yes || $this->questionHelper->ask($input, $output, $question)) {
+				$output->writeln("Decryption started\n");
+				$progress = new ProgressBar($output);
+				$progress->start();
+				$progress->setMessage("Decryption progress...");
+				$progress->advance();
+
+				$this->decryptAllUsersFiles($progress, $output);
+				$progress->finish();
+
+				if (empty($this->failed)) {
+
+					$this->IAppManager->disableApp('encryption');
+
+					//Delete the files_encryption dir
+					$this->rootView->deleteAll('files_encryption');
+
+					$this->appConfig->setValue('core', 'encryption_enabled', 'no');
+					$this->appConfig->deleteKey('encryption', 'useMasterKey');
+					$this->appConfig->deleteKey('encryption', 'masterKeyId');
+					$this->appConfig->deleteKey('encryption', 'recoveryKeyId');
+					$this->appConfig->deleteKey('encryption', 'publicShareKeyId');
+					$this->appConfig->deleteKey('files_encryption', 'installed_version');
+
+				}
+				$output->writeln("\nDecryption completed\n");
+
+				//Reencrypt again
+				$this->IAppManager->enableApp('encryption');
+				$this->appConfig->setValue('core', 'encryption_enabled', 'yes');
+				$this->appConfig->setValue('encryption', 'enabled', 'yes');
+				$output->writeln("Encryption started\n");
+
+				$output->writeln("Waiting for creating new masterkey\n");
+
+				$this->keyManager->setPublicShareKeyIDAndMasterKeyId();
+
+				$output->writeln("New masterkey created successfully\n");
+
+				$this->appConfig->setValue('encryption', 'enabled', 'yes');
+				$this->appConfig->setValue('encryption', 'useMasterKey', '1');
+
+				$this->keyManager->validateShareKey();
+				$this->keyManager->validateMasterKey();
+				$this->encryptAllUsers($input, $output);
+				$output->writeln("\nEncryption completed successfully\n");
+			} else {
+				$output->writeln("The process is abandoned");
+			}
+		} else {
+			$output->writeln("Master key is not enabled.\n");
+		}
+	}
+
+	protected function encryptAllUsers(InputInterface $input, OutputInterface $output) {
+		/*
+		 * We are reusing the encryptAll code but not the decryptAll. The reason being
```

Comment:
Added doc which says why we are not using decryptAll and re-using encryptAll. This should help future devs understand why we used this approach.